internal/xss/xss.go

```diff
@@ -2,188 +2,56 @@ package xss
 
 import (
 	"bytes"
-	"html/template"
-	"io/ioutil"
+	"github.com/Contrast-Security-OSS/go-test-bench/internal/common"
 	"log"
-	"net/http"
 	"net/url"
-	"strings"
-
-	"github.com/Contrast-Security-OSS/go-test-bench/internal/common"
 )
 
-func queryHandler(w http.ResponseWriter, r *http.Request, safety string) (template.HTML, bool) {
-	var input string
-
+// RegisterRoutes is to be called to add the routes in this package to common.AllRoutes.
+func RegisterRoutes(frameworkSinks ...*common.Sink) {
```

Comment:
It looks like you're basing this on `unvalidated redirect`. That one's a special case, and requires logic specific to each framework. It's possible to write this more like `path traversal`, using `VulnerableFnWrapper`:

```go
func RegisterRoutes() {
	sinks := []*common.Sink{
		{
			Name:     "generic",
			Sanitize: html.EscapeString,
			VulnerableFnWrapper: func(_ interface{}, payload string) (data string, raw bool, err error) {
				return payload, true, nil
			},
			RawMime: "text/html",
		},
	}
	common.Register(common.Route{
		// ...
```

Once you do that, most of the stuff outside of `RegisterRoutes` becomes unnecessary.

Note `RawMime`, which is currently not part of Sink. Without this, the mime type is set to `text/plain`, preventing the payload script from firing.

This patch adds RawMime:
```diff
diff --git a/internal/common/handler.go b/internal/common/handler.go
index aa67e3b..53428e7 100644
--- a/internal/common/handler.go
+++ b/internal/common/handler.go
@@ -41,6 +41,9 @@ func GenericHandler(s *Sink) (HandlerFn, error) {
                }
                }
                res, raw, err := s.VulnerableFnWrapper(opaque, payload)
                if raw {
+                       if len(s.RawMime) > 0 {
+                               mime = s.RawMime
+                       }
                        return res, mime, http.StatusOK
                }
 
diff --git a/internal/common/sink.go b/internal/common/sink.go
index 5f64053..13964d6 100644
--- a/internal/common/sink.go
+++ b/internal/common/sink.go
@@ -34,6 +34,10 @@ type Sink struct {
        // nil when this is set.
        VulnerableFnWrapper VulnerableFnWrapper
 
+       // the mime type used when VulnerableFnWrapper returns true for R1 (raw);
+       // defaults to text/plain.
+       RawMime string
+
        // http status that we expect to be returned for unsafe queries (used in testing)
        // defaults to http.StatusOK if unset
        ExpectedUnsafeStatus int
```

----

Once the above sink is defined, almost all that's needed for swagger to work is to regenerate the swagger stuff. I'd ask that you do that too, but it seems that something got out of sync with swagger, so it'd be less straightforward than it ought to be. I'll create a separate ticket to fix that up and add xss to swagger.